### PR TITLE
fix for linux hooks

### PR DESCRIPTION
**Pull request by pm100**
pre-commit now works

also improved the layout of error message dialog to preserve the message formatting

### CODE REVIEW

**Comment by pm100** [comment]
actually this fix works on windows too, but it requires git 4 windows to be installed (or anything else that puts bash on the devs path)

File: README.md [hunk 1]
```diff
@@ -36,7 +36,7 @@
 - Scalable terminal UI layout
 - Async [input polling](assets/perf_compare.jpg)
 - Async git API for fluid control
-
+a
```
Comment (extrawurst) [inline README.md:39]: ?

File: asyncgit/src/sync/hooks.rs [hunk 2]
```diff
@@ -94,13 +110,8 @@ fn run_hook(
     hook_script: &str,
     args: &[&str],
 ) -> HookResult {
-    let mut bash_args = vec![hook_script.to_string()];
```
Comment (extrawurst) [inline asyncgit/src/sync/hooks.rs:97]: why not just prepend the "-c" here?

File: src/components/commit.rs [hunk 3]
```diff
@@ -194,6 +194,16 @@ impl CommitComponent {
     }
 
     fn commit_msg(&mut self, msg: String) -> Result<()> {
+        if let HookResult::NotOk(e) = sync::hooks_pre_commit(CWD)? {
+            log::error!("pre-commit hook error: {}", e);
+            self.queue.borrow_mut().push_back(
+                InternalEvent::ShowErrorMsg(format!(
+                    "commit-msg hook error:\n{}",
```
Comment (extrawurst) [inline src/components/commit.rs:201]: I think this needs to be `pre-commit` right?

**Review by extrawurst — CHANGES_REQUESTED** [review]

File: asyncgit/src/sync/hooks.rs (hunk 2 above)
Comment (pm100) [reply asyncgit/src/sync/hooks.rs:97]: because its a 2 item vec not one string. I pass a vec of 2 args "-c" and "'script' arg1 arg2 arg3..."

Comment (pm100) [reply asyncgit/src/sync/hooks.rs:97]: the prepend -c is what I first did, didnt work the args didnt get passed to the script right

File: src/components/msg.rs [hunk 4]
```diff
@@ -32,17 +32,26 @@ impl DrawableComponent for MsgComponent {
...
-        );
 
-        let area = ui::centered_rect_absolute(65, 25, f.size());
+        // determine the maximum width of text block
+        let lens = self
+            .msg
+            .split('\n')
+            .map(str::len)
+            .collect::<Vec<usize>>();
+        let max = lens.iter().max().expect("max");
```
Comment (extrawurst) [inline src/components/msg.rs:42]: We never make sure that msg is not empty which would panic here, right?

File: asyncgit/src/sync/hooks.rs [hunk 5]
```diff
@@ -45,6 +46,21 @@ pub fn hooks_commit_msg(
     }
 }
 
+/// this hook is documented here https://git-scm.com/docs/githooks#_pre_commit
+///
+pub fn hooks_pre_commit(repo_path: &str) -> Result<HookResult> {
```
Comment (extrawurst) [inline asyncgit/src/sync/hooks.rs:51]: Can you add a test for this ?

File: src/components/msg.rs [hunk 6]
```diff
@@ -32,17 +32,26 @@ impl DrawableComponent for MsgComponent {
...
+        let mut width = u16::try_from(*max + 2).expect("cant fail");
```
Comment (extrawurst) [inline src/components/msg.rs:43]: Better `expect(“why it can’t fail”)`

File: src/components/msg.rs (hunk 4 above)
Comment (pm100) [reply src/components/msg.rs:42]: no its fine, just displays an empty box. (I tested it)

**Comment by pm100** [comment]
FYI - this code works on both linux and windows on my dev system. But I think thats because I have WSL installed. I am not sure what will happen on the github test systems. So I have tests for both windows and non windows. I will check that in and see what happens. I have resolved all yr other issues.

File: asyncgit/src/sync/hooks.rs [hunk 7]
```diff
@@ -204,6 +215,52 @@ exit 0
...
+sys.exit(0)
+        ";
+
+        create_hook(root, HOOK_PRE_COMMIT, hook);
+        let res = hooks_pre_commit(repo_path).unwrap();
+        assert_eq!(res, HookResult::Ok);
+    }
+
+    #[test]
+    fn test_pre_commit_fail_py() {
```
Comment (extrawurst) [inline asyncgit/src/sync/hooks.rs:242]: can you also add a shell based hook test

File: src/components/msg.rs [hunk 8]
```diff
@@ -32,17 +32,30 @@ impl DrawableComponent for MsgComponent {
...
-        );
 
-        let area = ui::centered_rect_absolute(65, 25, f.size());
+        // determine the maximum width of text block
+        let lens = self
+            .msg
+            .split('\n')
+            .map(str::len)
+            .collect::<Vec<usize>>();
+        let mut max = lens.iter().max().expect("max") + 2;
```
Comment (extrawurst) [inline src/components/msg.rs:42]: this will still panic if self.msg is empty

**Comment by extrawurst** [comment]
@pm100 two more things and we can merge this

File: src/components/msg.rs (hunk 8 above)
Comment (pm100) [reply src/components/msg.rs:42]: no it wont,  I have actually tested it and it works fine.

**Pull request merged**